src/codecoverage/backend/codecoverage_backend/services/gcp.py

```diff
@@ -0,0 +1,248 @@
+        '''
+        When a report exist for a changeset, download it and update redis data
+        '''
+        assert isinstance(push_id, int)
+
+        # Download the report
+        if not self.download_report(repository, changeset):
+            return
+
+        # Add the changeset to the sorted sets of known reports
+        # The numeric push_id is used as a score to keep the ingested
+        # changesets ordered
+        self.redis.zadd(KEY_REPORTS.format(repository=repository), {changeset: push_id})
+
+        logger.info('Ingested report', changeset=changeset)
+        return True
+
+    def download_report(self, repository, changeset):
+        '''
+        Download and extract a json+zstd covdir report
+        '''
+        # Chek the report is available on remote storage
+        path = '{}/{}.json.zstd'.format(repository, changeset)
+        blob = self.bucket.blob(path)
+        if not blob.exists():
+            logger.debug('No report found on GCP', path=path)
+            return False
+
+        archive_path = os.path.join(self.reports_dir, blob.name)
+        json_path = os.path.join(self.reports_dir, blob.name.rstrip('.zstd'))
+        if os.path.exists(json_path):
+            logger.info('Report already available', path=json_path)
+            return True
+
+        os.makedirs(os.path.dirname(archive_path), exist_ok=True)
+        blob.download_to_filename(archive_path)
+        logger.info('Downloaded report archive', path=archive_path)
+
+        with open(json_path, 'wb') as output:
+            with open(archive_path, 'rb') as archive:
+                dctx = zstd.ZstdDecompressor()
+                reader = dctx.stream_reader(archive)
+                while True:
+                    chunk = reader.read(16384)
+                    if not chunk:
+                        break
+                    output.write(chunk)
+
+        logger.info('Decompressed report', path=json_path)
+        return True
+
+    def store_push(self, repository, push_id, changesets):
+        '''
+        Store a push on redis cache, with its changesets
+        '''
+        assert isinstance(push_id, int)
+        assert isinstance(changesets, list)
+
+        # Store changesets initial data
+        for changeset in changesets:
+            key = KEY_CHANGESET.format(
+                repository=repository,
+                changeset=changeset,
+            )
+            self.redis.hset(key, 'push', push_id)
+
+        logger.info('Stored new push data', push_id=push_id)
+
+    def find_report(self, repository, min_push_id=None, max_push_id=None):
+        '''
+        Find the first report available before that push
+        '''
+        results = self.list_reports(repository, nb=1, min_push_id=min_push_id, max_push_id=max_push_id)
+        if not results:
+            raise Exception('No report found')
+        return results[0]
+
+    def find_closest_report(self, repository, revision):
```

Comment:
Nit: we might want to keep a consistent naming and either always use `changeset` or `revision`